gst/mqtt/ntputil.c

```diff
@@ -0,0 +1,289 @@
+ * |                                                               |
+ * +                      Receive Timestamp (64)                   +
+ * |                                                               |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |                                                               |
+ * +                      Transmit Timestamp (64)                  +
+ * |                                                               |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |                                                               |
+ * .                                                               .
+ * .                    Extension Field 1 (variable)               .
+ * .                                                               .
+ * |                                                               |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |                                                               |
+ * .                                                               .
+ * .                    Extension Field 2 (variable)               .
+ * .                                                               .
+ * |                                                               |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |                          Key Identifier                       |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |                                                               |
+ * |                            dgst (128)                         |
+ * |                                                               |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ *******************************************************************
+ */
+
+/**
+ * @brief A custom data type to represent NTP packet header format
+ */
+typedef struct _ntp_packet_t
+{
+  uint8_t li_vn_mode;
+  uint8_t stratum;
+  uint8_t poll;
+  uint8_t precision;
+  uint32_t root_delay;
+  uint32_t root_dispersion;
+  uint32_t ref_id;
+  ntp_timestamp_t ref_ts;
+  ntp_timestamp_t org_ts;
+  ntp_timestamp_t recv_ts;
+  ntp_timestamp_t xmit_ts;
+} ntp_packet_t;
+
+const uint64_t NTPUTIL_TIMESTAMP_DELTA = 2208988800ULL;
+const double NTPUTIL_MAX_FRAC_DOUBLE = 4294967295.0L;
+const int64_t NTPUTIL_SEC_TO_USEC_MULTIPLIER = 1000000;
+const char NTPUTIL_DEFAULT_HNAME[] = "pool.ntp.org";
+const uint16_t NTPUTIL_DEFAULT_PORT = 123;
+static pthread_once_t run_once_block = PTHREAD_ONCE_INIT;
+static pthread_spinlock_t lock;
+static int64_t local_monotonic_usec = -1;
+static int64_t ntp_timestamp_usec = -1;
+static int64_t ntp_poll_interval_usec = -1;
+
+/**
+ * @brief Initialize the NTP utility library using pthread_once
+ */
+void
+ntputil_init (void)
+{
+  pthread_spin_init (&lock, PTHREAD_PROCESS_SHARED);
+}
+
```

Comment:
No need to add some api for `pthread_spin_destroy` ?

Reply:
@anyj0527 Actually we need to call ```pthread_spin_destory```, but since the ```lock``` is statically allocated and it has the same life cycle as that of this library, this can be omitted (there is no point to invoke ```pthread_spin_destroy```).

Reply:
Right, It could be omitted. Thanks!